daemon/cmd/local_node_init.go

```diff
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	k8sLabels "k8s.io/apimachinery/pkg/labels"
+
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/k8s"
+	"github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/node/addressing"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/source"
+)
+
+type localNodeInitializerParams struct {
+	cell.In
+
+	Config    *option.DaemonConfig
+	Clientset client.Clientset
+	LocalNode k8s.LocalNodeResource
+}
+
+// localNodeInitializer performs the bootstrapping of the LocalNodeStore,
+// which contains information about the local Cilium node populated from
+// configuration and Kubernetes.
+type localNodeInitializer struct {
+	localNodeInitializerParams
+}
+
+func (ini *localNodeInitializer) InitLocalNode(ctx context.Context, n *node.LocalNode) error {
+	n.Source = source.Local
+
+	if err := ini.initFromConfig(ctx, n); err != nil {
+		return err
+	}
+
+	if err := ini.initFromK8s(ctx, n); err != nil {
```

Comment:
It seems like this aspect is pulling a k8s dependency in much earlier, or node initialization much later in the initialization process. From the commit message that seems somewhat intentional. I guess this raises the same question I had earlier, how do we reason about the initialization ordering?

Related, I see that newDaemon has the steps for node discovery and `k8s.waitForNodeInformation(), how do those aspects line up with this initializer?

Reply:
The `WaitForNodeInformation` was kept as-is except for the node IP, name and labels as those are not dependent on configuration or IPAM mode. Point is to resolve those much earlier so that we can lift more things out from daemon into modules. This makes the initialization safer.

Reply:
Is it fair to say that "By moving more of this logic behind an Observable Cell, this commit makes progress towards a code structure where the full node object may be initialized before exposing any updates for the node to the other subsystems. This provides stronger guardrails against race conditions where users of the Node objects may inadvertently access the structure before it has been initialized"?

I'm think the core of my question is "why is it safe to move this initialization earlier, will it cause other logic to be reordered?". If I understand, the counter-argument is "well it's not safe today, other modules could attempt to read local node state before it's initialized". So by moving towards a better structure, maybe at some point we can move all of the users of the local node over to relying on the `Observable` and then we will be able to eliminate this entire class of bugs. :+1: 

Then my concern about reordering other logic should be a non-issue, the new `initFromK8s()` logic is just doing a k8s fetch + updating some fields in what's effectively an internal structure. That's not inherently changing any ordering of other logic or triggering any events, it's just partially initializing a structure. So otherwise the current ordering of operations should not change, some data is just available earlier. I guess the one thing that does change here is that now we rely on the watcher for the local node to synchronize with k8s apiserver before continuing startup. That logic is already filtering the watch for only the current node, so it should be relatively quick, so no concerns there.

Reply:
I've now made this a bit more safer by leaving the `localNode` global variable used by the getters and setters unset by default and setting it from the `LocalNodeStore` start hook. This allows catching reordering bugs early, e.g. if we'd  modularize something out of daemon which uses e.g. `node.GetIPv4` instead of `LocalNodeStore` it would crash early and thus would give a hint to the author to switch to using `LocalNodeStore`.